Universal Music services code

lawrenceakka opened the pull request:
Continuing the discussion from #247 , with a PR!

This implements the MusicService class for access to all the music services known to Sonos.

It does not (yet) translate the data returned by the various services to DIDL.

Please test with the various services you have access to, and let me know what you think.

Lawrence

Examples:

Print all the services Sonos knows about

```python
>>> from soco.music_services import MusicService
>>> from pprint import pprint
>>> print (MusicService.get_all_music_services())
['Spotify',
 'The Hype Machine',
 'Saavn',
 'Bandcamp',
 'Stitcher SmartRadio',
 'Concert Vault',
 ...
     ]
```
 Or just those those to which you are subscribed (and the relevant usernames)

```python
>>> pprint (MusicService.get_subscribed_music_services())
[('Spotify', u'XXXXXX'), ('radioPup', u''), ('Spreaker', u'')]
```
Interact with TuneIn

```python

>>> tunein = MusicService('TuneIn')
>>> print tunein
<MusicService 'TuneIn' at 0x10ad84e10>
>>> pprint (tunein.get_metadata(item_id='root'))
{'count': 5,
 'index': 0,
 'mediaCollection': ({'albumArtURI': u'http://spotify-...',
                      'canEnumerate': True,
                      'canPlay': False,
                      'canScroll': False,
                      'id': u'playlists',
                      'itemType': u'favorites',
                      'title': u'Playlists'},
                     {'albumArtURI': u'http://spotify-...',
                      'canEnumerate': True,
                      'canPlay': True,
                      'canScroll': False,
                      'id': u'starred',
                      'itemType': u'favorites',
                      'title': u'Starred'},
                    ...),
 'total': 5}
# Get some metadata about a track
>>> response =  spotify.get_media_metadata(
... item_id='spotify:track:6NmXV4o6bmp704aPGyTVVG')
>>> print (response)
{'mediaMetadata': [{'id': u'spotify:track:6NmXV4o6bmp704aPGyTVVG'},
                   {'itemType': u'track'},
                   {'title': u'B?n Fra Helvete (Live)'},
                   {'mimeType': u'audio/x-spotify'},
                   {'trackMetadata': {'album': u'Mann Mot Mann (Ep)',
                                      'albumArtURI': u'http://o.s...9',
                                      'albumId': u'spotify:album:...',
                                      'artist': u'Kaizers Orchestra',
                                      'artistId': u'spotify:artist...',
                                      'canAddToFavorites': True,
                                      'canPlay': True,
                                      'canSkip': True,
                                      'duration': 317}}]}
```

or even a playlist

```python


>>> response =  spotify.get_metadata(
...    item_id='spotify:user:spotify:playlist:0FQk6BADgIIYd3yTLCThjg')

```
or a URI

```python


>>> response =  spotify.get_media_URI(
...    item_id='spotify:track:6NmXV4o6bmp704aPGyTVVG')

```
Find the available search categories, and use them

```python

>>> pprint (spotify.available_search_categories)
['albums', 'tracks', 'artists']
>>> result =  spotify.search(category='artists', term='miles')

```

**Test Result Table**

Music Service  | Instantiate | avail. search cats | search | metadata | media meta | media uri
------------------ | -------------- | ---------------------- | -------- | ------------ | --------------- | ------------
Deezer      | OK | OK | OK | OK | OK | OK
WiMP        | OK | OK | OK | OK    | OK | OK
YouSee Play | OK | OK | OK | OK | OK | OK
Google Play Music | OK | OK | OK | OK | OK | MIXED

KennethNielsen commented:
Hi @lawrenceakka 

I was planning to do some test on this, but I afraid I'm stopped before I even got started. The ``get_subscribed_music_services`` method does not work for me, and since that is also used in ``__init__``, I cannot instantiate music services. I tracked the issue down to my speakers not reporting the relevant information on the 'http://{Player_IP}:1400/status/securesettings' page. If I wget that page, from any one of the speakers, I just get:

```xml
 <?xml version="1.0" ?>
 <?xml-stylesheet type="text/xsl" href="/xml/review.xsl"?><ZPSupportInfo type="User"></ZPSupportInfo>
```

which has no content in the 'ZPSupportInfo' tag, (so in line 269 we try to run XML.fromstring on None). Maybe this was changed with a recent update or it is somehow settings dependent. I think I could work around it in __init__, but I wanted to get your input on it first.

lawrenceakka commented:
@KennethNielsen It looks as though this was changed in the most recent firmware update, probably to accommodate multiple user accounts for each service. I have rebased and updated, so it should work now.

lawrenceakka commented:
I've taken this out of Milestone 0.11.  There is still some work to do on it. For example, the recent firmware update means that there can be multiple accounts for the same music service on the system.

There are a few other changes I want to make as well.

KennethNielsen commented:
Hallo @lawrenceakka 

I tested the music services I have available to me and I created a results table in you initial PR post (after fighting with github over tables). As you can see most of it went fine.

The only thing I have to note I am really unsure of whether it is an error.
YouSee Play and Deezer fails on `get_metadata` when the item_id points to a track. But I think I have observed this before, that the dehaviour when trying to browse the end points of the content tree differs.

YouSee Play fails with the traceback:

```python
Traceback (most recent call last):
  File "test_ms.py", line 93, in test_music_service
    musicservice.get_metadata(item_id=item['id'])
  File "/home/kenneth/code/virt_env/soco/SoCo/soco/music_services/music_service.py", line 393, in get_metadata
    headers=self.headers
  File "/media/deepspace9/code/virt_env/soco/local/lib/python2.7/site-packages/pysimplesoap/client.py", line 264, in call
    raise SoapFault(unicode(response.faultcode), unicode(response.faultstring))
SoapFault: s:Server.ServiceUnknownError: An unknown service error occurred.
```

and Deezer with the traceback:

```python
Traceback (most recent call last):
  File "test_ms.py", line 93, in test_music_service
    musicservice.get_metadata(item_id=item['id'])
  File "/home/kenneth/code/virt_env/soco/SoCo/soco/music_services/music_service.py", line 393, in get_metadata
    headers=self.headers
  File "/media/deepspace9/code/virt_env/soco/local/lib/python2.7/site-packages/pysimplesoap/client.py", line 264, in call
    raise SoapFault(unicode(response.faultcode), unicode(response.faultstring))
SoapFault: s:Client: Node does not support browsing
```

the only question is what we should do about it, if anything. I don't think it is customary for libraries to allow exceptions from 3rd party libraries to pass though, so it should probably be handled. The question then is whether to raise a custom SoCo exception or simply return an empty search result. But since the error messages we get are not the same, nor do both of them contain enough information, I would say that the only real choice is to raise a custom SoCo exception.

Let me know of there is anything else that I can do.

KennethNielsen commented:
For everyone else, I have created a quick and dirty test script with this gist: https://gist.github.com/KennethNielsen/d28e635b77e2370e6a19

With that you should essentially be able to just check out this branch and run the script to be able to fill out the test table for your music services.

KennethNielsen commented:
Actually, probably as expected, the get_media_metadata and get_media_url also fails when trying to call them on ids that belong to containers. They also raise pysimplesoaps ``SoapFault``, which we should probably also catch and replace.

lawrenceakka commented:
I have done a fair bit of work on this, and the API has changed slightly, so I'm afraid I broke @KennethNielsen's script.  We now cater for multiple accounts per service (introduced by firmware 5.2)

I have wrapped the SoapFaults so that a MusicServicesException is raised instead, which is a bit more friendly.

zapp42 commented:
I tried the updated script from  lawrenceakka but I am stuck. First of all discovery() returns an empty set most of time so I "fixed it up" like this:
diff --git a/soco/music_services/music_service.py b/soco/music_services/music_service.py
index afbba90..ebdb3d4 100644
--- a/soco/music_services/music_service.py
+++ b/soco/music_services/music_service.py
@@ -91,7 +91,8 @@ class MusicAccount(object):
         # via a call to
         # systemProperties.GetStringX([('VariableName','R_SvcAccounts')]))
         # This returns an encrypted string, and, so far, we cannot decrypt it
-        device = soco or SoCo.any_soco()
+        #device = soco or SoCo.any_soco()
+        device = SoCo("192.168.178.22")
         log.debug("Fetching account data from %s", device)
         settings_url = "http://{0}:1400/status/accounts".format(
             device.ip_address)
@@ -472,7 +473,8 @@ class MusicService(object):
         credentials_header = result.add_child("credentials")
         credentials_header['xmlns'] = "http://www.sonos.com/Services/1.1"

-        device = SoCo.any_soco()
+        #device = SoCo.any_soco()
+        device = SoCo("192.168.178.22")
         device_id = device.systemProperties.GetString(
             [('VariableName', 'R_TrialZPSerial')])['StringValue']
         credentials_header.marshall('deviceId', device_id)


Now I can run the script but I am unable to instantiate a service:
all_services OK
get_subscribed_music_services OK

Instantiate Amazon Music FAILED

Instantiate Google Play Music FAILED

Instantiate Spotify FAILED

### SUMMARY
all_services................................................:OK
get_subscribed_music_services...............................:OK

Instantiate Amazon Music....................................:FAILED

Instantiate Google Play Music...............................:FAILED

Instantiate Spotify.........................................:FAILED

Any idea?

lawrenceakka commented:
@zapp42 

Please can you try something simpler, and report what happens

```python

>>> from soco.music_services import MusicService
>>> from pprint import pprint
>>> pprint (MusicService.get_all_music_services_names())
>>> spotify = MusicService('Spotify')
>>> pprint (spotify.get_metadata())

```

zapp42 commented:
@lawrenceakka 
Here you go:

>>> from soco.music_services import MusicService
>>> from pprint import pprint
>>> pprint (MusicService.get_all_music_services_names())
['7digital',
 'Calm Radio',
 'Bandcamp',
 'simfy',
 'hotelradio.fm',
 'Spreaker',
 'Hearts of Space',
 'Rdio',
 'Deezer',
 'Shuffler.fm',
 'Tribe of Noise',
 'RUSC',
 'SoundCloud',
 'radioPup',
 'Saavn',
 'Concert Vault',
 'MTV Music powered by Napster',
 'Live365',
 'Daytrotter',
 'DAR.fm Record Radio',
 'Murfie',
 'The Hype Machine',
 'Stitcher SmartRadio',
 'Qobuz',
 'Radionomy',
 'WiMP',
 'Juke',
 'MLB.com Gameday Audio',
 'Amazon Music',
 'Google Play Music',
 'Spotify',
 'TuneIn',
 'MixRadio',
 '8tracks']
>>> spotify = MusicService('Spotify')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "soco/music_services/music_service.py", line 340, in __init__
    self.headers = self._get_headers()
  File "soco/music_services/music_service.py", line 485, in _get_headers
    ('Username', self.account.username)
  File "soco/services.py", line 159, in _dispatcher
    return self.send_command(action, *args, **kwargs)
  File "soco/services.py", line 352, in send_command
    self.handle_upnp_error(response.text)
  File "soco/services.py", line 412, in handle_upnp_error
    error_xml=xml_error
soco.exceptions.SoCoUPnPException: UPnP Error 1000 received:  from 192.168.178.22

lawrenceakka commented:
Hmmm ... tests are currently failing on Python 2.6 (only). No idea why

zapp42 commented:
For completeness, this is what I get w/o any modifications from my side:
>>> spotify = MusicService('Spotify')            
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "soco/music_services/music_service.py", line 319, in __init__
    for acct in MusicAccount.get_all_accounts():
  File "soco/music_services/music_service.py", line 169, in get_all_accounts
    a for a in cls.get_account_data().values() if not a.deleted
  File "soco/music_services/music_service.py", line 118, in get_account_data
    root = XML.fromstring(cls._get_account_xml())
  File "soco/music_services/music_service.py", line 95, in _get_account_xml
    device = soco or SoCo.any_soco()
  File "soco/core.py", line 227, in any_soco
    device = discover().pop()
KeyError: 'pop from an empty set'

lawrenceakka commented:
@zapp42 
> soco.exceptions.SoCoUPnPException: UPnP Error 1000 received: from 192.168.178.22

UPnP Error 1000 is a network timeout, indicating routing or communication problems on your network:  see e.g. http://forums.sonos.com/showthread.php?p=93764

This is also why discovery is not working.

Do you have a firewall on your computer? What happens if you turn it off.

Please can you open a new issue on Github for this (cross refer to this one), as we are getting a bit off-topic now!

Lawrence

deekue commented:
I tried out Google Play Music with your code and I get error 8009.  I've search high and low but can't find a reference to that error number.

>>> gm = MusicService('Google Play Music')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "soco/music_services/music_service.py", line 340, in __init__
    self.headers = self._get_headers()
  File "soco/music_services/music_service.py", line 489, in _get_headers
    ('Username', self.account.username)
  File "soco/services.py", line 155, in _dispatcher
    return self.send_command(action, *args, **kwargs)
  File "soco/services.py", line 356, in send_command
    self.handle_upnp_error(response.text)
  File "soco/services.py", line 416, in handle_upnp_error
    error_xml=xml_error
soco.exceptions.SoCoUPnPException: UPnP Error 8009 received:  from 192.168.136.230

lawrenceakka commented:
@deekue 

> I tried out Google Play Music with your code and I get error 8009. I've search high and low but can't find a reference to that error number.

It looks like Google Play is using different authentication, which I have not implemented. I will fix it, probably tomorrow - I have some code mostly working

lawrenceakka commented:
I have done some fairly major refactoring (again), and some detailed wiresharking and proxying. Google Play now works. I hate Python 2.6 just at the moment.

You need to ensure that you have the very latest pysimplesoap, direct from github.

```pip install --upgrade --no-cache-dir -r requirements.txt``` should do it.

deekue commented:
there doesn't seem to be a version 1.17 for pysimplesoap.  the zip file the URL points to in requirements.txt returns version 1.16

lawrenceakka commented:
@deekue, 

Don't worry about it.  There is no such thing as v1.17. The zip file should actually contain the latest development code, but the author has not updated the version strings.  Install it and you should be fine

deekue commented:
ok, installed.

I now get this
```
>>> from soco.music_services import MusicService
>>> gm = MusicService('Google Play Music')
>>> gm.search(category='artists', term='miles')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "soco/music_services/music_service.py", line 448, in search
    ('count', count))
  File "soco/music_services/soap_transport.py", line 79, in call
    method, *args, headers=headers, **kwargs
  File "/usr/local/lib/python2.7/dist-packages/pysimplesoap/client.py", line 268, in call
    operation = self.get_operation(method)
  File "/usr/local/lib/python2.7/dist-packages/pysimplesoap/client.py", line 325, in get_operation
    for service_name, service in self.services.items():
AttributeError: 'NoneType' object has no attribute 'items'
```

lawrenceakka commented:
@deekue 

You're now hitting a bug in pysimplesoap that I discovered a few days ago.  See here: https://github.com/pysimplesoap/pysimplesoap/pull/38

Until that is fixed, are you able to apply that patch manually to your installed copy?  Or alternatively, uninstall pysimplesoap again (!) and install it instead from my fork at 
https://github.com/lawrenceakka/pysimplesoap/archive/fault-no-wsdl.zip

```pip install --upgrade https://github.com/lawrenceakka/pysimplesoap/archive/fault-no-wsdl.zip#egg=PySimpleSOAP-1.17 ```

should do it.

deekue commented:
installed your fork and now I'm getting results.  anything you'd like me to test?

lawrenceakka commented:
Thanks @deekue 

At the moment, try running @KennethNielsen 's test code (revised) at https://gist.github.com/lawrenceakka/4264c59da463fff0bc93/ and let us know how you get on.

Lawrence

deekue commented:
$ python gistfile1.py
all_services OK
get_subscribed_music_services OK

<MusicService 'Google Play Music' at 0x7f98a7e4edd0>
Instantiate Google Play Music OK
Google Play Music get_metadata root OK
Google Play Music get_metadata on items in root OK
Google Play Music available_search_categories OK
Google Play Music === search(albums) OK
Google Play Music results in search OK
Google Play Music get_metadata on albums OK
Google Play Music === search(tracks) OK
Google Play Music results in search OK
Google Play Music get_metadata on tracks FAILED
Traceback (most recent call last):
  File "gistfile1.py", line 93, in test_music_service
    musicservice.get_metadata(item_id=item['id'])
  File "/home/apathy/src/
github.com/lawrenceakka/SoCo/soco/music_services/music_service.py", line
416, in get_metadata
    ('count', count), ('recursive', recursive)
  File "/home/apathy/src/
github.com/lawrenceakka/SoCo/soco/music_services/soap_transport.py", line
106, in call
    raise MusicServiceException(exc.faultstring, exc.faultcode)
MusicServiceException: (u'Invalid metadata ID in request.', u'soap:Client')

Google Play Music get_media_metadata on tracks OK
Google Play Music get_media_uri on tracks FAILED
Google Play Music === search(artists) OK
Google Play Music results in search OK
Google Play Music get_metadata on artists OK

### SUMMARY
all_services................................................:OK
get_subscribed_music_services...............................:OK

Instantiate Google Play Music...............................:OK
Google Play Music get_metadata root.........................:OK
Google Play Music get_metadata on items in root.............:OK
Google Play Music available_search_categories...............:OK
Google Play Music === search(albums)........................:OK
Google Play Music results in search.........................:OK
Google Play Music get_metadata on albums....................:OK
Google Play Music === search(tracks)........................:OK
Google Play Music results in search.........................:OK
Google Play Music get_metadata on tracks....................:FAILED
Google Play Music get_media_metadata on tracks..............:OK
Google Play Music get_media_uri on tracks...................:FAILED
Google Play Music === search(artists).......................:OK
Google Play Music results in search.........................:OK
Google Play Music get_metadata on artists...................:OK


On Mon, Feb 9, 2015 at 9:28 AM, Lawrence Akka <notifications@github.com>
wrote:

> Thanks @deekue <https://github.com/deekue>
>
> At the moment, try running @KennethNielsen
> <https://github.com/KennethNielsen> 's test code (revised) at
> https://gist.github.com/lawrenceakka/4264c59da463fff0bc93/ and let us
> know how you get on.
>
> Lawrence
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SoCo/SoCo/pull/262#issuecomment-73551009>.
>



-- 
regards,
Daniel Quinlan

KennethNielsen commented:
@lawrenceakka Great work with this. I installed the patched pysimplysoap, [changed the test script](https://gist.github.com/KennethNielsen/cba1a14b8a83f44cab3e) (from your version, small change around line 96) to assume that it is ok to raise MusicServiceException when getting metadata for tracks and its all flying.

First of, I'm now only getting the exception with YouMusic and the test script now marks it as an acceptable exception and the other two services test all fine. I've update my results in the table:

### SUMMARY
all_services................................................:OK
get_subscribed_music_services...............................:OK

Instantiate YouSee Play Musik...............................:OK
YouSee Play Musik get_metadata root.........................:OK
YouSee Play Musik get_metadata on items in root.............:OK
YouSee Play Musik available_search_categories...............:OK
YouSee Play Musik === search(albums)........................:OK
YouSee Play Musik results in search.........................:OK
YouSee Play Musik get_metadata on albums....................:OK
YouSee Play Musik === search(playlists).....................:OK
YouSee Play Musik results in search.........................:OK
YouSee Play Musik get_metadata on playlists.................:OK
YouSee Play Musik === search(tracks)........................:OK
YouSee Play Musik results in search.........................:OK
YouSee Play Musik get_metadata on tracks....................:ACCEPTABLE EXCEPTION
YouSee Play Musik get_media_metadata on tracks..............:OK
YouSee Play Musik get_media_uri on tracks...................:OK
YouSee Play Musik === search(artists).......................:OK
YouSee Play Musik results in search.........................:OK
YouSee Play Musik get_metadata on artists...................:OK

Instantiate WiMP............................................:OK
WiMP get_metadata root......................................:OK
WiMP get_metadata on items in root..........................:OK
WiMP available_search_categories............................:OK
WiMP === search(albums).....................................:OK
WiMP results in search......................................:OK
WiMP get_metadata on albums.................................:OK
WiMP === search(playlists)..................................:OK
WiMP results in search......................................:OK
WiMP get_metadata on playlists..............................:OK
WiMP === search(tracks).....................................:OK
WiMP results in search......................................:OK
WiMP get_metadata on tracks.................................:OK
WiMP get_media_metadata on tracks...........................:OK
WiMP get_media_uri on tracks................................:OK
WiMP === search(artists)....................................:OK
WiMP results in search......................................:OK
WiMP get_metadata on artists................................:OK

Instantiate Deezer..........................................:OK
Deezer get_metadata root....................................:OK
Deezer get_metadata on items in root........................:OK
Deezer available_search_categories..........................:OK
Deezer === search(albums)...................................:OK
Deezer results in search....................................:OK
Deezer get_metadata on albums...............................:OK
Deezer === search(playlists)................................:OK
Deezer results in search....................................:OK
Deezer get_metadata on playlists............................:OK
Deezer === search(tracks)...................................:OK
Deezer results in search....................................:OK
Deezer get_metadata on tracks...............................:OK
Deezer get_media_metadata on tracks.........................:OK
Deezer get_media_uri on tracks..............................:OK
Deezer === search(artists)..................................:OK
Deezer results in search....................................:OK
Deezer get_metadata on artists..............................:OK

KennethNielsen commented:
I updated the results table to include @deekue results. The failed metadata should be ok, since it is for a track and it raises MusicServiceException, but the failed result on get URI will have to be looked into.

zapp42 commented:
Hi All. I have updated the music_services to the latest version (incl. the patched simplesoap stuff), but I still get upnp error 1000 for Amazon and Spotify and 8009 for Rdio in the GetSessionId call. 
My Sonos works w/o problems otherwise, so I really wonder why it works for you but not for me.

Can anybody tell me, how to trigger the music service related messages in the Sonos Controller app, so I can capture them with WS?

zapp42 commented:
Update:

I have reauthorized all my services in the Sonos Controller App and now I do not get upnp errors anymore. Instead I get the following (after a small change to the test script (adding try/except around the test_music_service() call):

rolf@NAS1701:~/git/SoCo$ python cba1a14b8a83f44cab3e/test_ms.py
all_services OK
get_subscribed_music_services OK

<MusicService 'Amazon Music' at 0xa728b0>
Instantiate Amazon Music OK
Amazon Music get_metadata root OK
Amazon Music get_metadata on items in root FAILED
Amazon Music available_search_categories OK
Amazon Music === search(albums) FAILED
Amazon Music === search(tracks) FAILED
Amazon Music === search(artists) FAILED

<MusicService 'Google Play Music' at 0xa8aef0>
Instantiate Google Play Music OK
Traceback (most recent call last):
  File "cba1a14b8a83f44cab3e/test_ms.py", line 136, in <module>
    test_music_service(service)
  File "cba1a14b8a83f44cab3e/test_ms.py", line 42, in test_music_service
    root = musicservice.get_metadata(item_id='root')
  File "/home/rolf/git/SoCo/soco/music_services/music_service.py", line 416, in get_metadata
    ('count', count), ('recursive', recursive)
  File "/home/rolf/git/SoCo/soco/music_services/soap_transport.py", line 79, in call
    method, *args, headers=headers, **kwargs
  File "/usr/local/lib/python2.7/dist-packages/pysimplesoap/client.py", line 268, in call
    operation = self.get_operation(method)
  File "/usr/local/lib/python2.7/dist-packages/pysimplesoap/client.py", line 325, in get_operation
    for service_name, service in self.services.items():
AttributeError: 'NoneType' object has no attribute 'items'


<MusicService 'Spotify' at 0xa82bf0>
Instantiate Spotify OK
Traceback (most recent call last):
  File "cba1a14b8a83f44cab3e/test_ms.py", line 136, in <module>
    test_music_service(service)
  File "cba1a14b8a83f44cab3e/test_ms.py", line 42, in test_music_service
    root = musicservice.get_metadata(item_id='root')
  File "/home/rolf/git/SoCo/soco/music_services/music_service.py", line 416, in get_metadata
    ('count', count), ('recursive', recursive)
  File "/home/rolf/git/SoCo/soco/music_services/soap_transport.py", line 79, in call
    method, *args, headers=headers, **kwargs
  File "/usr/local/lib/python2.7/dist-packages/pysimplesoap/client.py", line 268, in call
    operation = self.get_operation(method)
  File "/usr/local/lib/python2.7/dist-packages/pysimplesoap/client.py", line 325, in get_operation
    for service_name, service in self.services.items():
AttributeError: 'NoneType' object has no attribute 'items'


### SUMMARY
all_services................................................:OK
get_subscribed_music_services...............................:OK

Instantiate Amazon Music....................................:OK
Amazon Music get_metadata root..............................:OK
Amazon Music get_metadata on items in root..................:FAILED
Amazon Music available_search_categories....................:OK
Amazon Music === search(albums).............................:FAILED
Amazon Music === search(tracks).............................:FAILED
Amazon Music === search(artists)............................:FAILED

Instantiate Google Play Music...............................:OK

Instantiate Spotify.........................................:OK

Note: I now have OADevIds in http://playerip/status/accounts. This field was empty before reauthorizing all services. Still, the services worked fine.

zapp42 commented:
Update2:

Results after *really* using the patched pysimplesoap library:
rolf@NAS1701:~/git/SoCo$ python cba1a14b8a83f44cab3e/test_ms.py
all_services OK
get_subscribed_music_services OK

<MusicService 'Amazon Music' at 0x1134a90>
Instantiate Amazon Music OK
Amazon Music get_metadata root OK
Amazon Music get_metadata on items in root FAILED
Amazon Music available_search_categories OK
Amazon Music === search(albums) FAILED
Amazon Music === search(tracks) FAILED
Amazon Music === search(artists) FAILED

<MusicService 'Google Play Music' at 0x112bf50>
Instantiate Google Play Music OK
Google Play Music get_metadata root OK
Google Play Music get_metadata on items in root OK
Google Play Music available_search_categories OK
Google Play Music === search(albums) OK
Google Play Music results in search OK
Google Play Music get_metadata on albums OK
Google Play Music === search(tracks) OK
Google Play Music results in search OK
Google Play Music get_metadata on tracks ACCEPTABLE EXCEPTION
Google Play Music get_media_metadata on tracks OK
Google Play Music get_media_uri on tracks FAILED
Google Play Music === search(artists) OK
Google Play Music results in search OK
Google Play Music get_metadata on artists OK

<MusicService 'Spotify' at 0x114cf70>
Instantiate Spotify OK
Spotify get_metadata root OK
Spotify get_metadata on items in root FAILED
Spotify available_search_categories OK
Spotify === search(albums) OK
Spotify results in search OK
Spotify get_metadata on albums OK
Spotify === search(tracks) OK
Spotify results in search OK
Spotify get_metadata on tracks ACCEPTABLE EXCEPTION
Spotify get_media_metadata on tracks OK
Spotify get_media_uri on tracks OK
Spotify === search(artists) OK
Spotify results in search OK
Spotify get_metadata on artists OK

### SUMMARY
all_services................................................:OK
get_subscribed_music_services...............................:OK

Instantiate Amazon Music....................................:OK
Amazon Music get_metadata root..............................:OK
Amazon Music get_metadata on items in root..................:FAILED
Amazon Music available_search_categories....................:OK
Amazon Music === search(albums).............................:FAILED
Amazon Music === search(tracks).............................:FAILED
Amazon Music === search(artists)............................:FAILED

Instantiate Google Play Music...............................:OK
Google Play Music get_metadata root.........................:OK
Google Play Music get_metadata on items in root.............:OK
Google Play Music available_search_categories...............:OK
Google Play Music === search(albums)........................:OK
Google Play Music results in search.........................:OK
Google Play Music get_metadata on albums....................:OK
Google Play Music === search(tracks)........................:OK
Google Play Music results in search.........................:OK
Google Play Music get_metadata on tracks....................:ACCEPTABLE EXCEPTION
Google Play Music get_media_metadata on tracks..............:OK
Google Play Music get_media_uri on tracks...................:FAILED
Google Play Music === search(artists).......................:OK
Google Play Music results in search.........................:OK
Google Play Music get_metadata on artists...................:OK

Instantiate Spotify.........................................:OK
Spotify get_metadata root...................................:OK
Spotify get_metadata on items in root.......................:FAILED
Spotify available_search_categories.........................:OK
Spotify === search(albums)..................................:OK
Spotify results in search...................................:OK
Spotify get_metadata on albums..............................:OK
Spotify === search(tracks)..................................:OK
Spotify results in search...................................:OK
Spotify get_metadata on tracks..............................:ACCEPTABLE EXCEPTION
Spotify get_media_metadata on tracks........................:OK
Spotify get_media_uri on tracks.............................:OK
Spotify === search(artists).................................:OK
Spotify results in search...................................:OK
Spotify get_metadata on artists.............................:OK

lawrenceakka commented:
OK, I think is ready to be committed (at long last). pysimplesoap is no longer a dependency, because it had bugs which caused serious problems (and were not being fixed). All metadata is now returned in nested OrderedDicts (a backport for py2.6 is available).

I think we are now at the stage where this should be rolled into the master branch, and we can then fix any bugs as they arise.

Please leave any comments, and I will commit in a day or so.

Lawrence

KennethNielsen commented:
@lawrenceakka We had some problems with it not working on all service earlier. But even if that is still the case, I agree that we should get it pulled and then we can sort out support for all services later. I will try to put some of the testing from this thread, along with some random walks down through the content tree, into a more organized dev tool.

In any case, like to see it go in, but I would like an opportunity to have another look at it, especially with adding a soap client and everything, but I don't have time today. Can you hold of on pulling for a day or so?

lawrenceakka commented:
@KennethNielsen , yes of course. Just say if you have any problems, and I'll take a look

Lawrence

KennethNielsen commented on setup.py:
```diff
@@ -66,12 +65,11 @@ def run_tests(self):
-    import importlib
-except ImportError:
-    REQUIREMENTS.append('importlib')
+REQUIREMENTS = ['requests', 'dicttoxml']
```
Is there a reason for not reading them from the file. There is always the risk of getting out of sync!

lawrenceakka replied:
This is left over from when I used pysimplesoap (which could not be read from requirements.txt because it had to be installed from git). I will fix this.  [But in fact, we are doing this the wrong way anyway. The requirements should be listed in setup.py, and pip will get them from there.  It is requirements.txt that needs editing - see https://caremad.io/2013/07/setup-vs-requirement/.  But that is for another PR!

KennethNielsen commented on setup.py:
```diff
@@ -66,12 +65,11 @@ def run_tests(self):
+# Python 2.6 does not have importlib or OrderedDicts, so we need to install
+# the backports
+if sys.version_info < (2,7):
+    REQUIREMENTS.extend(['importlib', 'ordereddict'])
```
If I don't remember wring, ordereddict was also one of the types in python-future, so yet another reason

lawrenceakka replied:
Yes, you are right.

KennethNielsen commented on soco/core.py:
```diff
@@ -238,6 +244,20 @@ def __str__(self):
         return '{0}("{1}")'.format(self.__class__.__name__, self.ip_address)
 
+    @classmethod
+    def any_soco(cls):
```
Not extremely important here, but do you think maybe this would fit better under discover. I know that I have many times also wanted to have a from_name method in discover, to just always get Kitchen. So I thought maybe it would make sense to have a few helpers there under discover.

lawrenceakka replied:
Excellent point.  I'll move it

lawrenceakka replied:
Actually, it's not that easy, since it keys into the inheritance mechanism in the SoCo class (which we need to support different subclasses of SoCo for people who want them).  The discovery module does not know enough about all of this.  I'll think about how we might do it, but for the moment, I might leave this where it is

KennethNielsen commented on soco/music_services/accounts.py:
```diff
@@ -0,0 +1,181 @@
+
+import requests
+
+from soco.core import SoCo
```
I think we use relative imports most other places and maybe we should think about enforcing it in the future.

lawrenceakka replied:
Yes, good idea.  We should use from __future__ import absolute_imports everywhere, and then use relative imports for solo.  I'll change this ref.

KennethNielsen commented on soco/music_services/accounts.py:
```diff
@@ -0,0 +1,181 @@
+    _all_accounts = weakref.WeakValueDictionary()
+
+    def __init__(self):
+        """Constructor.
```
Kind of nit-picking here, but it isn't really a constructor ;)

lawrenceakka replied:
:wink: changed

KennethNielsen commented on soco/music_services/accounts.py:
```diff
@@ -0,0 +1,181 @@
+        for account in accounts:
+            serial_number = account.get('SerialNum')
+            is_deleted = True if account.get('Deleted') == '1' else False
+            if cls._all_accounts.get(serial_number):
```
Took me a few read-throughs to figure out lines 142-153, maybe a short comment.

Maybe also write out account as the variable name

KennethNielsen commented on soco/music_services/music_service.py:
```diff
@@ -0,0 +1,780 @@
+import requests
+
+from soco import SoCo
+from soco.xml import XML
```
After reading up on absolute vs. relative a bit, it seems that it is still not a clear recommendation for one or the other yet. So I guess maybe never mind about this and the other comment until we figure out what we prefer.

KennethNielsen commented on soco/music_services/music_service.py:
```diff
@@ -0,0 +1,780 @@
+    """
+
+    def __init__(self, endpoint, timeout, music_service):
+        """ Constructor
```
Again, not really

KennethNielsen commented on soco/music_services/music_service.py:
```diff
@@ -0,0 +1,780 @@
+        # a genuine Sonos value. Google Play music seems to want this. The
+        # firmware release number (after 'Sonos/" is obviously fake, and
+        # higher than current values, in case Google only offers certain
+        # services to certain firmware rel eases. ") #
```
If I remember correctly, this is one of the values we can get from one of the info versions. As long as the check on the other end, does not check for actual versions, this is of course fine. If they ever do, maybe we should think about filling it in with an detected value.

lawrenceakka replied:
I can't find it in info, or from any UPnP calls.  If we can get it easily, I agree it makes sense to use a real value.

KennethNielsen replied:
Ahh. Maybe I remember wrong. I just thought that was one of the info items in the new patch for speaker_info. But you reviewed that, so probably not.

KennethNielsen commented on soco/music_services/music_service.py:
```diff
@@ -0,0 +1,780 @@
+            login_token = XML.Element('loginToken')
+            token = XML.SubElement(login_token, 'token')
+            token.text = music_service.account.oa_device_id
+            key = XML.SubElement(login_token, 'key')
```
I must admit that everytime I look at XML generation code I feel torn. On the one hand, I really think we should use the library to create them, instead of doing by hand, to get rid of excaping issues etc. On the other hand it always annoys me, that it gets to long winded. Every time we add a subelement is two lines and a use-once-and-throw-away variable. I sometimes think that maybe it would be better, to do something like:
```python
XML.SubElement(login_token, 'token').text = music_service.account.oa_device_id
```
although, it is a bit harder to read. I don't know, as I said. Torn.

KennethNielsen commented on soco/music_services/music_service.py:
```diff
@@ -0,0 +1,780 @@
+            http_headers=self.http_headers,
+            soap_action="http://www.sonos.com/Services/1"
+                        ".1#{0}".format(method),
+            soap_header=self.get_soap_header(),
```
From the read through I still not completely sure if there is anything that changes over time in the soap head. Since we fetch it here every time we call a method on the server, would it be worth it/possible to cache the soap_header?

lawrenceakka replied:
Done

KennethNielsen commented on soco/music_services/music_service.py:
```diff
@@ -0,0 +1,780 @@
+
+        try:
+            result_elt = message.call()
+        except SoapFault as exc:
```
... and of course unset the cache here

KennethNielsen commented on soco/music_services/music_service.py:
```diff
@@ -0,0 +1,780 @@
+        result = parse(
+            XML.tostring(result_elt), process_namespaces=True,
+            namespaces={'http://www.sonos.com/Services/1.1': None}
+        ).values()[0]
```
I'm assuming, for this to be valid, that xmltodict.parse always returns an ordereddict, although I don't think that was completely obvious from the doc string. At first, I was wondering about why you introduced the dependence here on xmltodict, if the only thing you do is to pull a single value out, but I guess that result in itself contains lots of values that we would rather have in a dict.

lawrenceakka replied:
Yes. It is because the result contains lots of values. xmltodict is necessary, since the alternative is to return XML (complete with namespaces!) and let the user parse it, which seems very unfriendly, and inconsistent with what we do elsewhere

KennethNielsen commented on soco/music_services/music_service.py:
```diff
@@ -0,0 +1,780 @@
+         ...
+         ]
+
+        Or just those tho which you are subscribed
```
s/tho/who/

KennethNielsen commented on soco/music_services/music_service.py:
```diff
@@ -0,0 +1,780 @@
+        returned.
+
+        >>> from json import dumps # Used for pretty printing ordereddicts
+        >>> print(dumps(tunein.get_metadata(), indent=4)
```
maybe use pprint in this example instead, as it is sort of an abuse of JSON ;)

lawrenceakka replied:
Unfortunately, `pprint` does not work on OrderedDicts.  This is noted as a python bug. This seems to be generally accepted workaround for the moment

KennethNielsen replied:
Ahh, noted.

KennethNielsen commented on soco/music_services/music_service.py:
```diff
@@ -0,0 +1,780 @@
+
+        >>> response =  spotify.get_media_metadata(
+        ... item_id='spotify:track:6NmXV4o6bmp704aPGyTVVG')
+        >>> print(dumps(response, indent=4)
```
as above

KennethNielsen commented on soco/music_services/music_service.py:
```diff
@@ -0,0 +1,780 @@
+    _music_services_data = None
+
+    def __init__(self, service_name, account=None):
+        """Constructor.
```
As earlier

KennethNielsen commented on soco/music_services/music_service.py:
```diff
@@ -0,0 +1,780 @@
+        """
+        # This is very inefficient - loops within loops within loops, and
+        # many network requests
+        # Optimise it?
```
But should be called rather infrequently. For now, I wouldn't bother. Let the performance become problem first.

KennethNielsen commented on soco/music_services/music_service.py:
```diff
@@ -0,0 +1,780 @@
+
+        Args:
+            item_id (str): The container or item to browse. Defaults to the
+                root item
```
In the docstring here it says "The container or item to BROWSE". Is this correct. Does this mean that this is the methods to use for browsing. I remember earlier when I looked at the implementation I think it is not quite clear what the different things that we can query for does/is.

lawrenceakka replied:
Yes, I think this is what is used for browsing. You can start by querying the root (or doing a search). You can then `get_metadata` on albums, which will return tracks, and you can then `get_metadata` on the individual tracks one you know their ids.

KennethNielsen commented on soco/music_services/music_service.py:
```diff
@@ -0,0 +1,780 @@
+        return response.get('searchResult', None)
+
+    def get_media_metadata(self, item_id):
+        """Get metadata for a media item.
```
This is sort of what I meant. This return "metadata" for an (media) item. But the ``get_metadata`` gets metadata for an item or a container. Is it the same, if it is for an item?

KennethNielsen commented on soco/music_services/music_service.py:
```diff
@@ -0,0 +1,780 @@
+        return response.get('getMediaURIResult', None)
+
+    def get_last_update(self):
+        """Get metadata for a container or item.
```
This must be a copy paste error?

lawrenceakka replied:
Oops

KennethNielsen commented on soco/music_services/music_service.py:
```diff
@@ -0,0 +1,780 @@
+        return response.get('getExtendedMetadataTextResult', None)
+
+
+def desc_from_uri(uri):
```
Awesome that you figured this out

lawrenceakka replied:
Took a bit of work!

KennethNielsen commented on soco/soap.py:
```diff
@@ -0,0 +1,263 @@
+# WSDL file, but contains various bugs which mean that we would have to use
+# a patched version (upstream releases are infrequent).  Since SONOS only
+# appears to use basic SOAP features, and after experimenting with other
+# libraries, it seems best to write our own.
```
This is sad but cudos

KennethNielsen commented on soco/soap.py:
```diff
@@ -0,0 +1,263 @@
+        """ Args:
+                endpoint (str): The SOAP endpoint URL for this client
+                method (str): The name of the method to call
+                parameters (list): A list of (name, value) tuples containing
```
I wonder, since this code is new and for now only used here, if we should default to taking parameters as an ordered dict (which in my view is the better match) and then change the rest of the code, when services is ported to use this. What do you think?

lawrenceakka replied:
Yes, we could use an ordereddict here, though you have to pass it a list of tuples in order to create it like this:

```OrderedDict([('param1': 1), ('param2': 2)])```

instead of 

``` OrderedDict({'Param1': 1, 'param2': 2}) ```

since the latter does not (necessarily) retain the right order.

So, I'm not sure it helps at all! (And this is why we ended up using a list of tuples in the services code).

KennethNielsen replied:
Ahh. I thought it was because we have no ordereddict in py 2.6

KennethNielsen commented on soco/soap.py:
```diff
@@ -0,0 +1,263 @@
+        """ Prepare the SOAP header for sending """
+
+        if soap_header is not None:
+            return '<s:Header>{0}</s:Header>'.format(soap_header)
```
Shouldn't we use ElemetTree for this. I always get a bit worried about escaping and such when creating XML by hand.

lawrenceakka replied:
It's not (easily) possible here since the `soap_header` parameter is a string representation of XML, not an Element itself.  We could use an Element for `soap_header` by changing the rest of the code, but I think it is probably not worth it on balance (at least as long as this works!!)

KennethNielsen commented on soco/soap.py:
```diff
@@ -0,0 +1,263 @@
+        for name, value in parameters:
+            # pylint: disable=bad-format-string
+            tag = "<{name}>{value}</{name}>".format(
+                name=name, value=escape("%s" % value, {'"': "&quot;"}))
```
As above about ElementTree

KennethNielsen commented:
Ok. I have left some in-line comment. This is all minor things. Generally I think this looks great. In my comment earlier, I didn't mean that I wanted to test it more, but just review it. I think this thread is already long enough, so even IF there are still problems, let's get it in and sort them out later.

I was wondering, do you have any thoughts about data structure support for this. As I remember, the type that the items report are all over the place for the different music services, which would make it difficult to create a static hierarchy. I was thinking about making just a base class and then a subclass factory that would return a new sub-class for each music service & type combincation. The baseclass would then:

 * Contain a reference to its service, so that it can fetch metadata lazily
 * Do some _gettattr__ magic to get metadata as needed, type cast when we are certain that we can (e.g. `is_playable`) and cache the properties
 * Be able to create its own metadata for use in add_to_queue and the likes

What is your thinking on this?

While talking about data services. I guess we definitely want to add those later, so maybe importing the music services should give a warning stating that this is a preliminary implementation and that the interface WILL change.

lawrenceakka commented:
@KennethNielsen 

Thanks very much for your comments. I *think* I have dealt with them all (but let me know if not). I will commit shortly and then we can deal with the inevitable problems! 

There is *some* sense to the structure of the data returned, but it is really quite complex:  see http://musicpartners.sonos.com/node/83 for examples.  You suggestion sounds like it might work, but on the other had it might add even more complication.  For the moment, lets get the code in and play with it, before deciding.

Lawrence

KennethNielsen commented:
@lawrenceakka about data structures. I see your point, but I wasn't really thinking anything as ambitious as fully mirroring the internal datastructure, but merely to to have search and browse return object that can again be browsed and can produce their own metadata so they can be added to the queue (which I guess I would sort of expect given how we have implemented the rest.

In any case, I agree, let's get it in and start playing with it and then see.
